src/smt/theory_str_noodler/util.cc

```diff
@@ -377,22 +377,133 @@ namespace smt::noodler::util {
+            nfa = conv_to_nfa_hex(to_app(child), m_util_s, m, alphabet);
+            for (const auto& initial : nfa.initial) {
+                // FIXME: Cannot that introduce errors if there are transitions leading to the initial states?
+                nfa.final.add(initial);
+            }
+        } else if (m_util_s.re.is_range(expr)) { // Handle range.
+            assert(false && "re.is_range(expr)");
+        } else if (m_util_s.re.is_reverse(expr)) { // Handle reverse.
+            assert(false && "re.is_reverse(expr)");
+        } else if (m_util_s.re.is_union(expr)) { // Handle union (= or; A|B).
+            SASSERT(expr->get_num_args() == 2);
+            const auto left{ expr->get_arg(0) };
+            const auto right{ expr->get_arg(1) };
+            SASSERT(is_app(left));
+            SASSERT(is_app(right));
+            nfa = Mata::Nfa::uni(conv_to_nfa_hex(to_app(left), m_util_s, m, alphabet),
+                                 conv_to_nfa_hex(to_app(right), m_util_s, m, alphabet));
+        } else if (m_util_s.re.is_star(expr)) { // Handle star iteration.
+            SASSERT(expr->get_num_args() == 1);
+            const auto child{ expr->get_arg(0) };
+            SASSERT(is_app(child));
+            nfa = conv_to_nfa_hex(to_app(child), m_util_s, m, alphabet);
+            for (const auto& final : nfa.final) {
+                for (const auto& initial : nfa.initial) {
+                    nfa.delta.add(final, Mata::Nfa::EPSILON, initial);
+                }
+            }
+            nfa.remove_epsilon();
+
+            // Make one initial final in order to accept empty string as is required by kleene-star.
+            SASSERT(!nfa.initial.empty());
+            nfa.final.add(*nfa.initial.begin());
+        } else if (m_util_s.re.is_plus(expr)) { // Handle positive iteration.
+            SASSERT(expr->get_num_args() == 1);
+            const auto child{ expr->get_arg(0) };
+            SASSERT(is_app(child));
+            nfa = conv_to_nfa_hex(to_app(child), m_util_s, m, alphabet);
+            for (const auto& final : nfa.final) {
+                for (const auto& initial : nfa.initial) {
+                    nfa.delta.add(final, Mata::Nfa::EPSILON, initial);
+                }
+            }
+            nfa.remove_epsilon();
+        } else if(m_util_s.str.is_string(expr)) { // Handle string literal.
+            SASSERT(expr->get_num_parameters() == 1);
+            nfa = create_word_nfa(expr->get_parameter(0).get_zstring());
+        } else if(is_variable(expr, m_util_s)) { // Handle variable.
+            assert(false && "is_variable(expr)");
+        }
+
+        // Whether to create complement of the final automaton.
         if (make_complement) {
             Mata::OnTheFlyAlphabet mata_alphabet{ Mata::Nfa::create_alphabet(nfa) };
             nfa = Mata::Nfa::complement(nfa, mata_alphabet);
         }
         return nfa;
     }
 
+    Nfa create_word_nfa(const zstring& word) {
```

Comment:
I think we could make it as a mata function at some point (different parameter though).

Reply:
ok